bin/dartdoc.dart

```diff
@@ -122,48 +123,69 @@ main(List<String> arguments) async {
     _printUsageAndExit(parser, exitCode: 1);
   }
 
+  final logJson = args['json'] as bool;
+
   // By default, get all log output at `progressLevel` or greater.
   // This allows us to capture progress events and print `...`.
   logging.Logger.root.level = progressLevel;
-  final stopwatch = new Stopwatch()..start();
-
-  // Used to track if we're printing `...` to show progress.
-  // Allows unified new-line tracking
-  var writingProgress = false;
-
-  logging.Logger.root.onRecord.listen((record) {
-    if (record.level == progressLevel) {
-      if (showProgress && stopwatch.elapsed.inMilliseconds > 250) {
-        writingProgress = true;
-        stdout.write('.');
-        stopwatch.reset();
+
+  if (logJson) {
+    logging.Logger.root.onRecord.listen((record) {
+      if (record.level == progressLevel) {
+        return;
       }
-      return;
-    }
 
-    stopwatch.reset();
-    if (writingProgress) {
-      // print a new line after progress dots...
-      print('');
-      writingProgress = false;
-    }
-    var message = record.message;
-    assert(message == message.trimRight());
-    assert(message.isNotEmpty);
-
-    if (record.level < logging.Level.WARNING) {
-      if (message.endsWith('...')) {
-        // Assume there may be more progress to print, so omit the trailing
-        // newline
-        writingProgress = true;
-        stdout.write(message);
+      var output = <String, dynamic>{'level': record.level.name};
+
+      if (record.object is Jsonable) {
+        output['data'] = record.object;
       } else {
-        print(message);
+        output['message'] = record.message;
       }
-    } else {
-      stderr.writeln(message);
-    }
-  });
+
+      print(JSON.encode(output));
```

Comment:
Writing this to stdout is OK if you can redirect stdout internally in dartdoc to something else when writing json, so an errant print statement somewhere doesn't break your JSON.

Indeed, that's how this should be implemented generally, assuming that's possible in dart.  That way you don't have to require that all output goes through this library.